[FIX] account: Allow multiple-company accounts to be copied.

antoine162 opened the pull request:
When we implemented the feature of company-shared accounts, we didn't let the user copy an account that belongs to multiple companies. The `copy()` method required a single company to be passed in the default, and the `create()` method could not handle multiple companies passed as `company_ids`.

The reasoning behind this was that a constraint requires all the companies in company_ids to have a code set, but the ORM can only set the code for a single company at account creation. Without a way of setting the code for multiple companies, you can't create a shared account.

This is really annoying, so we are changing it.

Now, the `create()` method takes a `code_by_company` argument in the `vals`. This is a mapping `{company_id: code}` which lets the caller specify the code the account should take in any number of companies. Of course, the code should be given for all companies in `company_ids`, otherwise the create will fail.

The `copy()` method takes advantage of this to create a new account with codes that are based on the copied account's, in each company. So you can now copy multi-company accounts.

taskid: 4119457

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/176966.png)](https://mergebot.odoo.com/odoo/odoo/pull/176966)

antoine162 commented:
(note for the reviewer: the task is not yet up-to-date unfortunately - I'll make sure it's clarified this afternoon)

hupo-odoo reviewed (commented):
I find your code very impressive! I haven't much to say on the code, but here is a few comments / questions I have regarding shared accounts not specifically related to this PR (I apologize if this is already solved in another PR):
 - The copy of an account works perfectly when it's linked to 2 companies but I have an issue when 3 companies share the same account, I have the error "The code must be set for every company to which this account belongs." which is raised, but haven't yet figured out the root cause.
 - In the form view of an account, is there a point to see the "Mapping" section even in single-company mode while it doesn't provide any information ?
 - For your information, when I create a new account and try to set multiple companies (in the list view), I get a traceback
 - Wouln't it be nice to refresh the page when merging accounts, such that the user clearly sees that the merged account is now gone ?

antoine162 commented on addons/account/models/account_account.py:
```diff
@@ -784,22 +784,32 @@ def copy_data(self, default=None):
+                companies_to_get_new_account_codes = companies if 'code' not in default else companies[1:]
+                vals['code_by_company'] = {}
+                for company in companies_to_get_new_account_codes:
+                    start_code = account.with_company(company).code or account.with_company(account.company_ids[0]).code
```
This could for example happen if the original account belonged, say, to a single company, but the `company_ids` in the `default` contained several companies. In that case, the account might not have a code defined for every company in the company_ids, in which case we just take one of the account codes defined for the account as a starting point.

antoine162 commented on addons/account/models/account_account.py:
```diff
@@ -902,7 +921,19 @@ def create(self, vals_list):
+                    if company_id != company.id:
+                        new_account.with_context({'allowed_company_ids': [company_id, company.id]}).code = code
+                if other_companies:
+                    new_account.write({'company_ids': [Command.link(c.id) for c in other_companies]})
```
I tend to do it just for variables in generator expressions or arguments of lambda functions, I personally think it's nice because it lets you know that the variable doesn't come from the surrounding scope. I've changed it as you suggested here (as I think the full variable name is very readable). Thanks!

antoine162 commented:
Thank you so much for your review!

> * The copy of an account works perfectly when it's linked to 2 companies but I have an issue when 3 companies share the same account, I have the error "The code must be set for every company to which this account belongs." which is raised, but haven't yet figured out the root cause.

You're right - I fixed that, and added a test case.

> * In the form view of an account, is there a point to see the "Mapping" section even in single-company mode while it doesn't provide any information ?

I guess there isn't, but I'll ask Benj and fix that in another PR, if needed.

> * For your information, when I create a new account and try to set multiple companies (in the list view), I get a traceback

That should also be fixed now.

> * Wouln't it be nice to refresh the page when merging accounts, such that the user clearly sees that the merged account is now gone ?

That will work seamlessly when we change the account merge to a wizard (https://github.com/odoo/odoo/pull/177296)

hupo-odoo commented on addons/account/models/account_account.py:
```diff
@@ -886,23 +887,32 @@ def name_create(self, name):
+                vals['company_ids'] = [Command.set(companies[0].ids)]
+            code_by_company_list = [vals.pop('code_by_company', {}) for vals in vals_list_for_company]
+            # Create the accounts with a single company and a single code.
+            new_accounts = super(AccountAccount, self.with_company(companies[0])).create(vals_list_for_company)
```
I get a traceback at this line of code when doing the following:
- have a shared account on multiple companies
- having only the first company selected in the company selector
- copy the account
=> Error

antoine162 replied:
Thanks :smile: it should be fixed now.
- The error was caused because we were setting fields with `check_company = True` when creating the account. Since at that point the account only has one company, this was breaking the check_company constraint.
- The solution is to defer the write of those fields to when we write all the `company_ids` on the account.
- It was an undeterministic error, due to the fact that we arbitrarily choose the first of the company_ids as the company we use to create the account at first.

hupo-odoo reviewed (commented):
Hello Antoine, thanks for your replies on all my comments/questions :smiley:. 
Just a suggestion, could you maybe give a little more spaces to your code, it's very compact, which make it sometimes harder to read

hupo-odoo commented:
LGTM, I'll transfer to my final reviewer :smile:
